[I2C] Multiple Instances (#3224)

TD-er opened the pull request:
See #3224

TD-er committed:
[I2C] multiple instances P047 - Environment - Soil moisture sensor

See #3224

TD-er committed:
[I2C] Multiple instances P079 - Wemos Motorshield

See #3224

TD-er committed:
[I2C] Multiple instances P012 - LCD 2004

TD-er committed:
[I2C] Multiple Instances P006 - BMP085/180

TD-er committed:
[I2C] Multiple instances P015 - TSL2561

TD-er committed:
[Cleanup] Cleanup P015

TD-er commented:
I've made a test build with these changes upto P015
[deleted]

Would be nice if those can already be tested to at least work as they did before, but even better when using multiple instances of them on the same node.
For some you cannot change the I2C address, so for those you really need to use an I2C multiplexer (which is also included in the build)

tonhuisman commented:
In what builds are the multiplexer included? I tried the `ESP_Easy_mega_20200820_test_ESP8266_4M1M_VCC` but the multiplexer isn't available. I'll try a `normal` build first.

TD-er commented:
Testing has `LIMIT_BUILD_SIZE` enabled, so that will exclude the multiplexer.
All other builds that have "normal" included and not using `LIMIT_BUILD_SIZE` will have the multiplexer code present.

tonhuisman commented:
`Normal` is getting me in the right direction, thanks. Found another BME280, so I can start some tests 😄

tonhuisman commented:
It took a while before the BME's woke up, but now they are working intermittently. After a while I see this in the log:
```
3014335 : Info  : BMx280 : Detected BME280
3014344 : Info  : BMx280 : Detected BME280
3014352 : Info  : BMx280 : Detected BME280
3015335 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3015336 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3015337 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3015338 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3015339 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3015340 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3016335 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3016336 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3016337 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3016337 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3016339 : Info  : BMx280 : Unable to detect chip ID (255, failed)
3016339 : Info  : BMx280 : Unable to detect chip ID (255, failed)
```

I'll try with a build without the BME changes, as it might be related to the way I've hooked them up, even though they come up just fine in the I2C scanner.

I marked the devices that are actually connected on my current setup (had a fourth TSL2561 laying around also).
![Screenshot - 21-08-2020 , 10_51_41](https://user-images.githubusercontent.com/1922393/90872282-b1772400-e39c-11ea-8f94-73fe5cf80d3c.png)

tonhuisman commented:
Aha, it has exactly the same errors in an older build, so I'll be hardware-debugging for a while 😞

tonhuisman commented:
It seems to help a bit to run this many devices on an ESP32, but the connection issues are the most probable cause for the above errors. And I don't want to solder it all together, as I intend to use these sensors with other ESP's. (I also swapped the ESP8266 for another model, still with 4MB flash, but the errors keep returning, so it must be connection issues, as wiggling the wires *does* make a difference 🦆)

TD-er commented:
Do you use pull-up resistors on the SDA/SCL lines?
Also notice any difference when using a lower I2C frequency?

tonhuisman commented:
I haven't used pull-ups yet, on any of the I2C devices, as it all 'just worked', but maybe it's now time to start using those (TCA954x documentation explicitly mentions that each channel should have pull-ups, but I presumed that most devices already had them incorporated, I think I'm wrong there 😄)

With BME's and TSL2561's I've not yet had any high frequency issues, so I don't think that'll help here. But it's worth a try.

TD-er commented:
Some of those "Chinese module boards" do use pull-up resistors, so it may depend on what combination of boards you use whether it may work.

On boards I design, I always at least include the pads for pull-up resistors, just to leave the option open for debugging.
But for 1 or 2 I2C modules I did not have to use them.
I guess it may differ per setup.

tonhuisman commented:
Yeah, it worked fine all week, but today, after re-wiring the sensors, it started acting up. I'll use a multi-meter to check if some do have pull-up (I suspect the OLED boards to have them, but I didn't connect any of them today, that could also be a factor)

tonhuisman commented:
Did some testing with pull-ups and lower frequency.
The TCA954x chips obviously, as documented, requires pull-ups on its input channel (SDA/SCL), so I soldered 2x 4k7 resistors on the board to solve that once and for all.
Running the I2C bus at 200 kHz does make the BME280 stable, and they have their own pull-ups (measured a consistent 10k between VCC and SDA/SCL on all BMS's I have here), but the TSL2561's probably hates the multiplexer, as they only run somewhat reliable at 100 kHz, either with 10 k or 4k7 pull-up resistors. Without pull-ups they now give nearly consistent I2C errors.
The OLED's don't have any pull-ups installed by default, but they seem to work reasonably well without.

tonhuisman commented:
Another possible cause related to `I2C errors` in common is that we have no concurrency-control implemented. This means of course that any plugin that needs to access it's hardware, sets the channel it needs, communicates with the device, and resets the multiplexer channel. I'll think (and read) about how to resolve that, as we don't want to implement complete blocking during communication (or maybe we do/should?). Setting the refresh/read interval to prime number values should help to avoid such clashes, but we'll need more testing, I guess, as the bus can get quite crowded.

TD-er commented:
See this issue: https://github.com/letscontrolit/ESPEasy/issues/3224
That's where I keep track of the plugins that should now be able to run with multiple instances of the same plugin.

The whole idea of this PR is just to add concurrency control.

tonhuisman commented:
But does that take care of resource-concurrency? As that seems timing-related, and not memory/object related.

TD-er commented:
> But does that take care of resource-concurrency? As that seems timing-related, and not memory/object related.

Well it does for "resources" as in task data that should be held in memory between plugin calls.

But maybe we need to add a delay after switching the I2C multiplexer port?  (is this mentioned in the datasheet?)
And as extra we could add some cache to keep track of the current active channels so you don't need to set the channel if it is the same as the last one.

TD-er committed:
Merge branch 'mega' into feature/I2C_multi_instance

TD-er committed:
[Cleanup] Mark floating point numbers with f to reduce resource usage

See: https://github.com/esp8266/Arduino/commit/953dfd945f03816db1943fa1345a9550fe61b422

tonhuisman commented on src/WebServer_AdvancedConfigPage.ino:
```diff
@@ -126,9 +126,9 @@ void handle_advanced() {
   addFormNumericBox(F("Timezone Offset (UTC +)"), F("timezone"), Settings.TimeZone, -720, 840); // UTC-12H ... UTC+14h
   addUnit(F("minutes"));
-  addFormFloatNumberBox(F("Latitude"), F("latitude"), Settings.Latitude, -90.0, 90.0);
+  addFormFloatNumberBox(F("Latitude"), F("latitude"), Settings.Latitude, -90.0, 90.0f);
```
Shouldn't the negative values also get the `f` marker?

TD-er replied:
Yep, missed those.
It is very likely I did miss a few more as I started with some query replace and then worked further on those files that had changes.

TD-er committed:
Add missing float marking

tonhuisman commented on src/_P092_DLbus.ino:
```diff
@@ -934,8 +934,8 @@ boolean P092_fetch_heatmeter(int number, sP092_ReadData* ReadData) {
   heat_meter = (DLbus_Data->ByteStream[HMindex.kwh_index + 1] << 8) | DLbus_Data->ByteStream[HMindex.kwh_index];
-  heat_meter_mwh = (heat_meter * 0.1) / 1000; // in MWh
-  if (heat_meter_mwh > 1.0) {
+  heat_meter_mwh = (heat_meter * 0.1f) / 1000f; // in MWh
```
The compiler doesn't appreciate that, expects 1000.0f or just 1000

TD-er replied:
Ah, had already fixed a number of those incorrect replaces I made.
Apparently I did not compile with this plugin.
Thanks for noticing.

TD-er committed:
Fix compiler issue with incorrect 1000f in P092 DLbus

TD-er committed:
[I2C multiplexer] Cleanup multiplexer code in _Plugin.ino

A lot of code duplication now moved to pre- and post- functions.

TD-er committed:
[I2C multiplexer] Further cleanup of code

TD-er committed:
Merge branch 'mega' into feature/I2C_multi_instance

TD-er committed:
Use new (std::nothrow)

TD-er committed:
Fix compiler warnings for LOLIN_EPD library

Also mentioned in this (unmerged) PR: https://github.com/wemos/LOLIN_EPD_Library/pull/7/files

TD-er committed:
[I2C] Multiple Instances P022 - PCA9685

TD-er committed:
[P022] Forgot #ifdef USES_P022

TD-er committed:
[I2C] Multiple Instances P023 OLED

TD-er committed:
[Cleanup] No need to call clearPluginTaskData from PLUGIN_EXIT 

It is already the default action on PLUGIN_EXIT

TD-er committed:
[I2C] Multiple Instances P024 MLX90614

TD-er committed:
[I2C] Multiple Instances P025 ADS1115

TD-er committed:
[I2C] Multiple Instances P027 INA219

TD-er committed:
[I2C] Move I2C.ino to helpers .h/.cpp

TD-er committed:
[I2C] Multiple Instances P032 MS5611

TD-er committed:
[Build] Fix build issues with LOLIN_EPD lib on minimal OTA builds

TD-er committed:
[I2C] Multiple instances P034 DHT12

TD-er committed:
[I2C] Multiple Instances P045 MPU6050

TD-er committed:
[I2C] Multiple Instances P048 Motorshield V2

TD-er committed:
[I2C] Multiple Instances P057 - Display - HT16K33 LED

TD-er committed:
[I2C] Multiple instances P058 - Keypad - HT16K33

TD-er committed:
[I2C] Multiple Instances P060 - Analog input - MCP3221

TD-er committed:
[Cleanup] Shorten member variable names

TD-er committed:
[I2C] Multiple Instances P062 - Keypad - MPR121 Touch

TD-er commented:
New test build.
- [ESPEasy_ESP32_mega-20200829-21-PR_3226.zip](https://www.dropbox.com/s/owp345i2sgw6b3j/ESPEasy_ESP32_mega-20200829-21-PR_3226.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200829-21-PR_3226.zip](https://www.dropbox.com/s/8rgm24muk3ukvp1/ESPEasy_ESP82xx_mega-20200829-21-PR_3226.zip?dl=0)

Changed plugins:

-  P006 - BMP085/180 I2C Barometric Pressure Sensor
-  P007 - Analog input - PCF8591
-  P009 - MCP23017 input (maybe not do this one, as it will also be part of the core GPIO structure)
-  P010 - Light/Lux - BH1750 (not needed, as it does not keep run time data)
-  P011 - Extra IO - ProMini Extender (not needed, as it does not keep run time data)
-  P012 - Display - LCD2004
-  P014 - Environment - SI7021/HTU21D
-  P015 - Light/Lux - TSL2561
-  P017 - RFID - PN532 (But may need some changes as it looks like blocking code instead of a state machine
-  P019 - PCF8574 input (maybe not do this one, as it will also be part of the core GPIO structure)
-  P022 - Extra IO - PCA9685
-  P023 - Display - OLED SSD1306
-  P024 - MLX90614 IR temperature I2C 0x5A)
-  P025 - ADS1115 I2C 0x48)
-  P027 - Energy (DC) - INA219
-  P028 - Environment - BMx280
-  P032 - MS5611 (GY-63) I2C Temp/Barometric Pressure Sensor
-  P034 - Temperature and Humidity sensor DHT 12 (I2C)
-  P036 - Display - OLED SSD1306/SH1106 Framed
-  P045 - Gyro - MPU 6050
-  P047 - Environment - Soil moisture sensor (very simple change as only the I2C address is kept as a member)
-  P048 - Motor - Adafruit Motorshield v2
-  P050 - Plugin 050 I2C TCS34725 RGB Color Sensor with IR filter and White LED
-  P051 - Plugin 051 Temperature and Humidity Sensor AM2320
-  P057 - Display - HT16K33 LED
-  P058 - Keypad - HT16K33
-  P060 - Analog input - MCP3221
-  P061 - PCF8574/MCP23017 KeyPad
-  P062 - Keypad - MPR121 Touch

TD-er committed:
[I2C] Multiple Instances  P064 - APDS9960 Gesture

TD-er committed:
[I2C] Multiple Instances P069 - LM75A Temperature Sensor (I2C)

Plus some duplicate includes in other files.

TD-er committed:
[I2C] Multiple instances P072 - Temp/Hum HDC1080 (no changes)

TD-er committed:
[I2C] Multiple instances P083 - SGP30 - Gas

TD-er committed:
[I2C] Multiple Instances P090 - CCS811 Air Quality TVOC/eCO2

TD-er committed:
[Cleanup] Reduce stack usage for P064 in web load function

TD-er commented:
Updated the last dropbox link with a new build.
It now also includes these:

-  P064 - APDS9960 Gesture
-  P066 - VEML6040 RGBW
-  P068 - Environment - SHT30/31/35
-  P069 - LM75A Temperature Sensor (I2C)
-  P072 - Temperature and Humidity sensor HDC1080 (I2C)
-  P074 - TSL2591 I2C Lux/IR Sensor
-  P079 - Wemos Motorshield (very simple change as only the I2C address is kept as a member)
-  P083 - SGP30 - Gas (TVOC - Total Volatile Organic Compounds) and Air Quality (eCO2)
-  P084 - VEML6070 UV
-  P090 - CCS811 Air Quality TVOC/eCO2 Sensor

tonhuisman commented:
I've tested a couple of sensors and a dual-OLED setup, but unfortunately didn't have enough time to try with multiple sensors and a multiplexer (though that is the main reason of this 'exercise'... 😮).

Most likely you have tested these sensors already, but all I could test was the APDS9960, BH1750, BME280, SSD1306 and SH1106.
All seems to work fine, in the short testing I was able to do. 👍 

There is 1 remark about the FramedOLED; the startup-logo seems to be displayed much shorter than before. I didn't check if a delay was removed from the code, or if it is just that the optimizations make it work faster, but the logo is replaced much quicker with the configured content. This isn't a *problem*, just something I noticed.

TD-er commented:
Thanks for testing.
I will also have a quick look at the code to see if I mixed up `!= nullptr` for `== nullptr` or vice verse.
If that checks out, I will merge it, so we can continue.
And hopefully the source will also be checked/tested by others before the next nightly build.

TD-er committed:
Merge branch 'mega' into feature/I2C_multi_instance

TD-er committed:
[Cleanup] Use enum class in P023 OLED

TD-er committed:
[Cleanup] Proper initialize all member variables of plugins

As reported by PVS-studio

TD-er committed:
[Cleanup] Uncrustify P092_DLbus

TD-er committed:
[Cleanup] Use enum class for P036_FramedOLED

TD-er committed:
[Cleanup] Rename member vars of P045 and initialize all members

TD-er committed:
[Cleanup] Rename member + initialize members for P090

TD-er committed:
Add check for AllocatedControllerSettings in storage

TD-er commented:
As it appeared the latest test builds may not have been using the actual changes, I made a new test build for this PR:

- [ESPEasy_ESP32_mega-20200829-82-PR_3226.zip](https://www.dropbox.com/s/rfxcdg3p8tobmf6/ESPEasy_ESP32_mega-20200829-82-PR_3226.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200829-82-PR_3226.zip](https://www.dropbox.com/s/hfmurxc3s457s72/ESPEasy_ESP82xx_mega-20200829-82-PR_3226.zip?dl=0)

TD-er committed:
Merge branch 'mega' into feature/I2C_multi_instance

PeterSturm commented:
Hallo TD-er,
the following download link does not work. It seems to have been removed.

ESPEasy_ESP82xx_mega-20200829-82-PR_3226.zip

Would you please be so kind and publish the link once again. That would be awesome.
Thanks a lot

TD-er commented:
This PR has been merged a while ago, so I don't have the PR builds anymore.
The code of this is now in the builds so you can just load any nightly build made after September 13 to have this functionality.

PeterSturm commented:
Ok, thanks a lot
